File: algorithms/Graphs/Medium/PrimsMST/VSDPrimsMST.py



# Function to find MST of given graph
def VSDprims(G,n):
    inf=999999
    visited=[0]*n
    edge_count=0
    visited[0]=True
    cost=0
    while(edge_count<n-1):
        minimum=inf
        x=0
        y=0
        for i in range(n):
            if visited[i]:
                for j in range(n):
                    if ((not visited[j]) and G[i][j]):
                        if minimum > G[i][j]:
                            minimum=G[i][j]
                            x=i
                            y=j
                            
        print("edge pair is ",(x,y)," weight is ",G[x][y])
        cost+=G[x][y]
        visited[y]=True
        edge_count+=1
    print("The total cost of Prim's MST is ",cost)


if __name__=="__main__":
    g =[ [ 0, 2, 0, 6, 0 ],
        [ 2, 0, 3, 8, 5 ],
        [ 0, 3, 0, 0, 7 ],
        [ 6, 8, 0, 0, 9 ],
        [ 0, 5, 7, 9, 0 ] 
    ]
    VSDprims(g,5)

File: algorithms/Graphs/Medium/PrimsMST/VSDPrimsMST.cpp

#include<bits/stdc++.h>


void prims(int graph[][5],int n){
    bool visited[n]{false};
    int edge_count=0;
    int cost=0;
    visited[0]=true;

    while(edge_count<n-1){
        int mini=INT_MAX;
        int x=0;
        int y=0;
        for(int i=0;i<n;i++){
            if(visited[i]){
                for(int j=0;j<n;j++){
                    if(!visited[j] && graph[i][j]!=0){
                        if(mini > graph[i][j]){
                            mini = graph[i][j];
                            x=i;
                            y=j;
                        }
                    }
                }
            }
        }
        std::cout<<"edge pair is "<<x<<" "<<y<<" "<<"weight is "<<graph[x][y]<<std::endl;
        cost+=graph[x][y];
        visited[y]=true;
        edge_count+=1;
    }
    std::cout<<"The total cost of Prim's MST is "<<cost;

}



int main(){
    int v=5;
    int graph[5][5] = { { 0, 2, 0, 6, 0 },
                    { 2, 0, 3, 8, 5 },
                    { 0, 3, 0, 0, 7 },
                    { 6, 8, 0, 0, 9 },
                    { 0, 5, 7, 9, 0 } 
                };
    prims(graph,v);
    return 0;
}

